### ISSUE TITLE
Weird Command Order Differences in SMTP TCP connection without TLS

### ISSUE THREAD

**Issue by MondoGao** [traefik/traefik#11302]
### Welcome!

- [x] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [x] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

As discussed here: https://github.com/stalwartlabs/mail-server/discussions/356#discussioncomment-11360578

I'm trying to proxy port 25 and 587 for a SMTP mail server. And traefik behaves differently with no proxy included situation.

Setting: 
- Proxy protocol disabled for port 25
- Stalwart implicit tls for 25 is disabled, and the port is exposed to different port, e.g. 8025
- Traefik entrypoint 25 -> router SNI * -> 8025, neither TLS nor TLS passthrough is enabled.

### What did you see instead?

Testing:
1. direct access to stalwart by telnet and port 25: without sending any command, we receive the welcom msg from stalwart server.
![CleanShot 2024-11-23 at 20 18 44](https://github.com/user-attachments/assets/c703013f-8059-4bb0-a340-eb85ad05ba60)
2. through traefik 25 -> 8025:  we can only receive the welcom msg after sending first packet (e.g. an invalid ehlo command as shown here)
![CleanShot 2024-11-23 at 20 22 14](https://github.com/user-attachments/assets/b19771e2-25bf-4f6c-aab6-6b8d8ea317e3)

I believe Traefik might be rejecting or buffering the initial welcome message from the mail server for some reason, and this affects all the mail clients I've tried so far.

The general recommendation for avoiding this issue is do not proxy any plain tcp (no tls and starttls) ports though traefix at all. And this is exactly what the above anwser suggests.

### What version of Traefik are you using?

3.2.1

### What is your environment & configuration?

```yaml
# docker-compose
services:
    container_name: stalwart-mail
    image: stalwartlabs/mail-server:latest
    ports:
      - 8025:25 # SMTP
    labels:
      - traefik.enable=true
      - traefik.tcp.routers.smtp.rule=HostSNI(`*`)
      - traefik.tcp.routers.smtp.entrypoints=smtp
      - traefik.tcp.routers.smtp.service=smtp@file
```

```yaml
# traefik.yml
entryPoints:
  smtp:
    address: :25
providers:
  docker:
  file:
    directory: ./config
```

```yaml
# config/stalwart.toml
[[tcp.services.smtp.loadBalancer.servers]]
address = "127.0.0.1:8025"
```

### If applicable, please paste the log output in DEBUG level

Noting special printed while conducting expr.

**Comment by nmengin** [comment]
Hey @MondoGao,

Thanks for reaching out.

If any community member can help us find verified steps to reproduce and fix the issue if possible, we would love the help.

**Comment by zone-zero** [comment]
I'm seeing the same issue with a postfix pod running plain SMTP on port 25 internally, and trying to expose it via port 25252 externally.


Ingress:
```YAML
---
apiVersion: traefik.io/v1alpha1
kind: IngressRouteTCP
metadata:
  name: mail-ingress-tcp
  namespace: notify

spec:
  entryPoints:
    - mail-tcp
  routes:
  - match: HostSNI(`*`)
    services:
    - name: mail
      port: 25
```
Traefik helmrelease via Flux (with other ports redacted)
```YAML
---
apiVersion: helm.toolkit.fluxcd.io/v2beta1
kind: HelmRelease
metadata:
  name: traefik
  namespace: ingress
spec:
  interval: 10m
  chart:
    spec:
      chart: ./traefik
      version: "~3.5.0"
      sourceRef:
        kind: GitRepository
        name: traefik
        namespace: flux-system
  values:
    installCRDs: true
    deployment:
      replicas: 1
    service:
      type: LoadBalancer
      annotations:
    ports:

   ...

      mail-tcp:
        port: 25252
        expose:
          default: true
        exposedPort: 25252
        protocol: TCP

   ...
    additionalArguments:
      - "--serversTransport.insecureSkipVerify=true" # ignore cert errors when communicating with containers
      - "--api.insecure=true" #expose api so it can be accessed by k8s services

    providers:
      kubernetesCRD:
        allowCrossNamespace: true

```

**Comment by zone-zero** [comment]
Seeing it with FTP as well.

**Comment by joejulian** [comment]
By the way... this isn't buffering the server-first message, it's not even establishing a connection until the client sends something.

**Comment by joejulian** [comment]
I figured it out.

The [documentation](https://doc.traefik.io/traefik/v3.2/routing/routers/#entrypoints_1) says:
> ❕ How to handle Server First protocols?
>
> To correctly handle a request, Traefik needs to wait for the first few bytes to arrive before it can decide what to do with it.
> 
> For protocols where the server is expected to send first, such as SMTP, if no specific setup is in place, we could end up in a situation where both sides are waiting for data and the connection appears to have hanged.
> 
> The only way that Traefik can deal with such a case, is to make sure that on the concerned entry point, there is no TLS router whatsoever (neither TCP nor HTTP), and there is at least one non-TLS TCP router that leads to the server in question.

Which seems kinda vague and not helpful. I had pretty-much the same configuration as @zone-zero and clearly have nothing TLS connected to that endpoint. I even checked in the dashboard and the TLS shield is not part of this router.

So time for some code digging.

Traefik’s TCP router has two execution paths:
- A direct forwarding path, where the backend connection is established immediately and server output is forwarded without reading from the client.
- A classification path, where Traefik attempts to determine whether the connection is TLS or plaintext by peeking at the client stream.

The direct forwarding path is taken only when:
- there is at least one non-TLS TCP router,
- there are no TLS TCP routers, and
- there are no HTTPS routers associated with the entrypoint.

https://github.com/traefik/traefik/blob/v3.6.7/pkg/server/router/tcp/router.go#L87-L118

If those conditions are not met, Traefik falls back to classification and peeks at the client stream:
```go
br.Peek(1)
```
This call blocks until the client sends at least one byte.

https://github.com/traefik/traefik/blob/v3.6.7/pkg/server/router/tcp/router.go#L370-L399

This explains why the SMTP greeting is only forwarded after sending an invalid command.

Even with a non-TLS TCP router (`HostSNI("*")`, and no `tls:` key), the classification path may still be selected if Traefik believes an HTTPS or TLS router exists on the same entrypoint.

One way this can happen is via [default entrypoint expansion](https://doc.traefik.io/traefik/reference/install-configuration/entrypoints/#asdefault): if no entrypoints are explicitly marked asDefault=true, Traefik treats all entrypoints as defaults, allowing HTTP routers without explicit entrypoint assignment to attach implicitly.

Once an HTTPS router is associated with the entrypoint, the direct forwarding path is bypassed.

TL;DR

Traefik forwards server-first TCP traffic immediately only when it can select a non-TLS TCP router without inspecting client bytes. If any HTTPS or TLS router is associated with the entrypoint (including implicitly via default entrypoints), Traefik blocks waiting for client data before forwarding, delaying the SMTP greeting.

**Comment by joejulian** [comment]
Is this a bug, or a documentation problem? I'm afraid it'll be considered documentation despite the principal of least surprise suggesting that this should be a bug - at least until the next major version allows for breaking changes.

**Comment by joejulian** [comment]
@nmengin It looks, to me, like breaking changes are allowed in main? If so I can also open a PR to make this behavior less surprising.

**Comment by nmengin** [comment]
Hey @joejulian,

Breaking changes are allowed on the main branch.
However, it means that the PR would not be evaluated before the next Major version, unless you provide a deprecation path.
In such a case, we could merge the feature in a minor version and delete the deprecated way in the next Major version.

**Comment by swinis** [comment]
Having the same issue on Port 25 with only one TCP Router (none TLS) bound to the entrypoint as "HostSNI('*') - No middleware or anything els configured. The Route is supposed to route to Stalwart Port 25. Traefik version is 3.6.10.

Any plans to take up this topic?
